### PR TITLE
feat: add stats for using vocamac

### PR DESCRIPTION
**Pull request by Mr-Sunglasses**
Now users can see the stats of how much words they transcribe using Vocamac, what's the words per minute, speed and what's the streak of them using Vocamac.

Screenshot 2026-06-01 at 3 44 34 PM

## AI and LLM's Assistance

I've used Gemini CLI to assist with this issue, but I personally reviewed all the generated changes before submitting the PR. I also added tests to cover the code introduced in this contribution.

**Model used:** Auto -- Gemini CLI selects the best model for each task automatically, choosing between [gemini-3.1-pro](https://deepmind.google/models/gemini/pro/) and [gemini-3-flash](https://deepmind.google/models/gemini/flash/).

### CODE REVIEW

**Review by jatinkrmalik — COMMENTED** [review]
@Mr-Sunglasses Thanks for this, the feature is well structured and the DI plumbing through `AppState`, `StatsManaging`, and `MockStatsManager` follows the existing service conventions closely. 

I have posted a couple of comments below. Please take a look at them.

File: Sources/VocaMac/Services/StatsManager.swift [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    func recordTranscription(_ transcription: VocaTranscription) {
+        let text = transcription.text.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !text.isEmpty else { return }
+
+        // Estimate word count
+        let words = text.components(separatedBy: .whitespacesAndNewlines)
+            .filter { !$0.isEmpty }
+            .count
+
+        let dateKey = formatDate(transcription.timestamp)
```
Comment (jatinkrmalik) [inline Sources/VocaMac/Services/StatsManager.swift]: `let dateKey = formatDate(transcription.timestamp)` keys daily buckets off `VocaTranscription.timestamp`, but that property is always set to `Date()` inside the model's `init` and the constructor argument is ignored. That is fine in production, but it is the root cause of the untestable streak path you flagged in the test file. Could we add a `timestamp` parameter to `VocaTranscription.init` (defaulting to `Date()`) so streak and daily-bucket logic becomes injectable and testable?

File: Sources/VocaMac/Services/StatsManager.swift [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            // First time usage
+            stats.currentStreak = 1
+            stats.bestStreak = 1
+            return
+        }
+
+        let calendar = Calendar.current
+
+        // Check if last usage was yesterday
+        if calendar.isDateInYesterday(lastDate) {
```
Comment (jatinkrmalik) [inline Sources/VocaMac/Services/StatsManager.swift:112]: The streak math relies on `Calendar.current` here while `formatDate` builds keys with a separate `DateFormatter` that has no timeZone. If those two ever disagree on day boundaries (user near midnight, or a TZ change), the daily bucket and the streak can diverge. Worth deriving both from one source of truth, ideally a single injected `Calendar`, which also makes `updateStreaks` unit-testable.

File: Sources/VocaMac/Services/StatsManager.swift [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            // Streak broken
+            stats.currentStreak = 1
+        }
+
+        if stats.currentStreak > stats.bestStreak {
+            stats.bestStreak = stats.currentStreak
+        }
+    }
+
+    private func formatDate(_ date: Date) -> String {
```
Comment (jatinkrmalik) [inline Sources/VocaMac/Services/StatsManager.swift]: `formatDate` builds a fresh `DateFormatter` on every call and sets no explicit `locale` or `timeZone`. Two issues: (1) `DateFormatter` is relatively expensive to allocate, and this runs on every transcription; (2) without `formatter.locale = Locale(identifier: "en_US_POSIX")` and an explicit `timeZone`, the `yyyy-MM-dd` keys can shift across locale/timezone changes, which would silently fragment a user's daily buckets and break streaks. Can we hoist a single `static let` formatter with POSIX locale and a fixed timeZone?

File: Sources/VocaMac/Models/UserStats.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    /// Daily word counts to calculate trends and streaks
+    /// Key is date string in "yyyy-MM-dd" format
+    var dailyWordCounts: [String: Int] = [:]
+
+    /// Daily duration to calculate WPM history
+    var dailyDurationSeconds: [String: Double] = [:]
+
+    /// Calculated average Words Per Minute (WPM)
+    var averageWPM: Double {
```
Comment (jatinkrmalik) [inline Sources/VocaMac/Models/UserStats.swift:36]: `averageWPM` divides total words by `totalAudioDurationSeconds`. Since `audioLengthSeconds` is the full clip length including leading and trailing silence, this systematically under-reports WPM versus what a user perceives as their speaking speed. That may be intentional, but could we add a `///` note clarifying it is words-per-minute-of-audio, not words-per-minute-of-speech, so the UI label is not misleading?

File: Sources/VocaMac/Views/StatsSettingsTab.swift [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+                                        Divider()
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    .padding(8)
+                }
+
+                Spacer()
```
Comment (jatinkrmalik) [inline Sources/VocaMac/Views/StatsSettingsTab.swift]: `Spacer()` inside a `VStack` that lives in a `ScrollView` is a no-op; a scroll view sizes to its content, so there is no extra vertical space for the spacer to push against. I would drop it, or use a fixed `.padding(.top, ...)` on the reset button if the intent was spacing.

File: Sources/VocaMac/Views/StatsSettingsTab.swift [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+                }
+                .controlSize(.small)
+                .frame(maxWidth: .infinity, alignment: .center)
+                .padding(.top, 20)
+            }
+            .padding()
+        }
+    }
+
+    private func formatDuration(_ seconds: Double) -> String {
```
Comment (jatinkrmalik) [inline Sources/VocaMac/Views/StatsSettingsTab.swift:154]: `formatDuration` duplicates a fair bit of manual seconds/minutes/hours math. `DateComponentsFormatter` with `unitsStyle = .abbreviated` and `allowedUnits = [.hour, .minute, .second]` gives the same "1h 2m" output, is localized, and removes the hand-rolled branches. Optional, but it would shrink this view and avoid edge cases.

File: Tests/VocaMacTests/StatsManagerTests.swift [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        // Record for yesterday
+        let t1 = VocaTranscription(text: "Yesterday", duration: 1.0, detectedLanguage: "en", audioLengthSeconds: 1.0, modelUsed: .tiny)
+        // Manually inject lastUsageDate to yesterday for testing
+        statsManager.recordTranscription(t1)
+        
+        // We need to simulate that the last usage was actually yesterday.
+        // Since recordTranscription sets it to now, we'll manually adjust it for the test.
+        // This requires making some properties internal or using a mockable date provider.
+        // For now, let's just test that it starts at 1.
+        XCTAssertEqual(statsManager.stats.currentStreak, 1)
```
Comment (jatinkrmalik) [inline Tests/VocaMacTests/StatsManagerTests.swift:84]: This test is named `testStreakIncrementsOnNewDay` but, as the inline comment admits, it cannot exercise the new-day path because `timestamp` is not injectable, so it only asserts `currentStreak == 1` on first use. That overlaps with `testInitialStatsAreEmpty` plus one record. Once `VocaTranscription` accepts a `timestamp` (see my StatsManager note), could we rewrite this to record yesterday then today and assert the streak becomes 2, and add a streak-broken case (gap of 2+ days resets to 1)? Those are the branches most likely to regress.

File: Tests/VocaMacTests/StatsManagerTests.swift [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+final class StatsManagerTests: XCTestCase {
+    var statsManager: StatsManager!
+    var cancellables: Set<AnyCancellable>!
+
+    @MainActor
+    override func setUp() {
+        super.setUp()
+        // Use a temporary file for testing persistence if needed, 
+        // but for logic tests we can just use a fresh instance.
+        statsManager = StatsManager()
```
Comment (jatinkrmalik) [inline Tests/VocaMacTests/StatsManagerTests.swift]: `setUp` does `StatsManager()` then `resetStats()`, but `StatsManager` reads and writes the real `stats.json` under Application Support. These tests mutate and overwrite the developer's (or CI's) actual stats file and depend on that shared on-disk state. Could we make the stats file location injectable (pass a URL or temp directory into `init`) so the suite runs against an isolated temp file? That also unblocks a real persistence round-trip test (record, reload from disk, assert).

File: Tests/VocaMacTests/Mocks/MockServices.swift [hunk 9]
```diff
@@ -327,6 +327,34 @@ final class MockTextInjector: TextInjecting {
...
+    @Published var stats: UserStats = UserStats()
+
+    var recordCallCount = 0
+    var resetCallCount = 0
+
+    var objectWillChangePublisher: AnyPublisher<Void, Never> {
+        objectWillChange.eraseToAnyPublisher()
+    }
+
+    func recordTranscription(_ transcription: VocaTranscription) {
```
Comment (jatinkrmalik) [inline Tests/VocaMacTests/Mocks/MockServices.swift:411]: Small drift: `MockStatsManager.recordTranscription` re-implements word counting independently of the real `StatsManager` and does not touch streaks or daily buckets. That is fine for a mock, but since `recordCallCount` is what the tests actually assert on, consider keeping the body minimal (increment counters only) so it does not imply it mirrors production logic.

**Review by jatinkrmalik — CHANGES_REQUESTED** [review]

**Review by jatinkrmalik — APPROVED** [review]
Thanks for the follow-up commits and for iterating on the stats implementation.
Appreciate the contribution, approving this and I will handle the date-boundary fix in a follow-up PR.
